onmt/inputters/inputter.py

```diff
@@ -209,306 +12,135 @@ class IterOnDevice(object):
-    )
-    merged_vocab = Vocab(
-        merged, specials=all_specials,
-        max_size=vocab_size, min_freq=min_freq
-    )
-    if vocab_size_multiple > 1:
-        _pad_vocab_to_multiple(merged_vocab, vocab_size_multiple)
-    src_field.vocab = merged_vocab
-    tgt_field.vocab = merged_vocab
-    assert len(src_field.vocab) == len(tgt_field.vocab)
-
-
-def _read_vocab_file(vocab_path, tag):
+        tgt_vocab = _read_vocab_file(opt.tgt_vocab,
+                                     opt.tgt_words_min_frequency)
+        tgt_vocab = pyonmttok.build_vocab_from_tokens(
+            tgt_vocab,
+            maximum_size=opt.tgt_vocab_size,
+            special_tokens=[DefaultTokens.UNK,
+                            DefaultTokens.PAD,
+                            DefaultTokens.BOS,
+                            DefaultTokens.EOS])
+        tgt_vocab.default_id = tgt_vocab[DefaultTokens.UNK]
+        vocabs['tgt'] = tgt_vocab
+
+    if opt.src_feats_vocab:
+        src_feats = {}
+        for feat_name, filepath in opt.src_feats_vocab.items():
+            src_f_vocab = _read_vocab_file(filepath, 1)
+            src_f_vocab = pyonmttok.build_vocab_from_tokens(
+                src_f_vocab,
+                maximum_size=0,
+                minimum_frequency=1,
+                special_tokens=[DefaultTokens.UNK,
+                                DefaultTokens.PAD,
+                                DefaultTokens.BOS,
+                                DefaultTokens.EOS])
+            src_f_vocab.default_id = src_f_vocab[DefaultTokens.UNK]
+            src_feats[feat_name] = src_f_vocab
+        vocabs['src_feats'] = src_feats
+
+    vocabs['data_task'] = opt.data_task
+
+    return vocabs
+
+
+def _read_vocab_file(vocab_path, min_count):
     """Loads a vocabulary from the given path.
 
     Args:
         vocab_path (str): Path to utf-8 text file containing vocabulary.
             Each token should be on a line, may followed with a count number
             seperate by space if `with_count`. No extra whitespace is allowed.
-        tag (str): Used for logging which vocab is being read.
     """
 
-    logger.info("Loading {} vocabulary from {}".format(tag, vocab_path))
-
     if not os.path.exists(vocab_path):
         raise RuntimeError(
-            "{} vocabulary not found at {}".format(tag, vocab_path))
+            "Vocabulary not found at {}".format(vocab_path))
     else:
         with codecs.open(vocab_path, 'r', 'utf-8') as f:
             lines = [line.strip() for line in f if line.strip()]
             first_line = lines[0].split(None, 1)
             has_count = (len(first_line) == 2 and first_line[-1].isdigit())
             if has_count:
-                vocab = [line.split(None, 1) for line in lines]
+                vocab = []
+                for line in lines:
+                    if int(line.split(None, 1)[1]) >= min_count:
+                        vocab.append(line.split(None, 1)[0])
```

Comment:
Not critical, but what's the rationale behind the maxsplit set to 1 here? This might do more harm than good IMO.

Reply:
was there before: https://github.com/OpenNMT/OpenNMT-py/blob/7e8c7c2a6ac1d28265f27c861fc63ef72c1f0fae/onmt/inputters/inputter.py#L511
I think it's in case the vocab file has more than two fields, but not really happening I guess.